crates/task-impls/src/transactions.rs

```diff
@@ -232,114 +232,102 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions> TransactionTask
-                    <TYPES::BlockPayload as BlockPayload<TYPES>>::Transaction,
-                > = Vec::new();
+        let mut sequencing_fees = Vec::new();
+        let mut transactions: Vec<<TYPES::BlockPayload as BlockPayload<TYPES>>::Transaction> =
+            Vec::new();
 
-                for bundle in bundles {
-                    sequencing_fees.push(bundle.sequencing_fee);
-                    transactions.extend(bundle.transactions);
-                }
+        for bundle in bundles {
+            sequencing_fees.push(bundle.sequencing_fee);
+            transactions.extend(bundle.transactions);
+        }
 
-                let validated_state = self.consensus.read().await.decided_state();
+        let validated_state = self.consensus.read().await.decided_state();
 
-                if let (Ok(sequencing_fees), Ok((block_payload, metadata))) = (
-                    Vec1::try_from_vec(sequencing_fees),
-                    TYPES::BlockPayload::from_transactions(
-                        transactions,
-                        &validated_state,
-                        &Arc::clone(&self.instance_state),
-                    )
-                    .await,
-                ) {
-                    broadcast_event(
-                        Arc::new(HotShotEvent::BlockRecv(PackedBundle::new(
-                            block_payload.encode(),
-                            metadata,
-                            block_view,
-                            sequencing_fees,
-                            None,
-                            Some(auction_result),
-                        ))),
-                        event_stream,
-                    )
-                    .await;
-
-                    return None;
-                }
-            } else {
-                warn!("Timeout while getting auction results");
-            }
-        }
+        let sequencing_fees = Vec1::try_from_vec(sequencing_fees)
+            .context("Failed to receive a bundle from any builder.")?;
+        let (block_payload, metadata) = TYPES::BlockPayload::from_transactions(
+            transactions,
+            &validated_state,
+            &Arc::clone(&self.instance_state),
+        )
+        .await?;
 
-        // If we couldn't get any bundles (due to either all of the builders or solver failing to return a result), send an empty block
-        warn!(
-            "Failed to get a block for view {:?}, proposing empty block",
-            block_view
-        );
+        Ok(PackedBundle::new(
+            block_payload.encode(),
+            metadata,
+            block_view,
+            sequencing_fees,
+            None,
+            Some(auction_result),
+        ))
+    }
 
+    /// Produce a null block
+    pub async fn null_block(
+        &mut self,
+        block_view: TYPES::Time,
+        version: Version,
+    ) -> Option<PackedBundle<TYPES>> {
         // Increment the metric for number of empty blocks proposed
```

Comment:
Nit: I'd make this function sync and move the metric increment to the caller to make this one side-effect free